relay/writer/file_util.go

```diff
@@ -0,0 +1,268 @@
+	if n == 0 && err == io.EOF {
+		return false, nil // empty file
+	} else if n != fileHeaderLen {
+		return false, errors.Errorf("binlog file %s has no enough data, only got % X", filename, buff[:n])
+	} else if err != nil {
+		return false, errors.Annotate(err, "read binlog header")
+	}
+
+	if !bytes.Equal(buff, replication.BinLogFileHeader) {
+		return false, errors.Errorf("binlog file %s header not valid, got % X, expect % X", filename, buff, replication.BinLogFileHeader)
+	}
+	return true, nil
+}
+
+// checkFormatDescriptionEventExist checks if the file has a valid FormatDescriptionEvent.
+func checkFormatDescriptionEventExist(filename string) (bool, error) {
+	// FormatDescriptionEvent always follows the binlog file header
+	exist, err := checkBinlogHeaderExist(filename)
+	if err != nil {
+		return false, errors.Annotatef(err, "check binlog file header for %s", filename)
+	} else if !exist {
+		return false, errors.Errorf("no binlog file header at the beginning for %s", filename)
+	}
+
+	f, err := os.Open(filename)
+	if err != nil {
+		return false, errors.Annotatef(err, "open file %s", filename)
+	}
+	defer f.Close()
+
+	// check whether only the file header
+	fileHeaderLen := len(replication.BinLogFileHeader)
+	fs, err := f.Stat()
+	if err != nil {
+		return false, errors.Errorf("get stat for %s", filename)
+	} else if fs.Size() == int64(fileHeaderLen) {
+		return false, nil // only the file header
+	}
+
+	// seek to the beginning of the FormatDescriptionEvent
+	_, err = f.Seek(int64(fileHeaderLen), io.SeekStart)
+	if err != nil {
+		return false, errors.Annotatef(err, "seek to %d for %s", fileHeaderLen, filename)
+	}
+
+	// parse a FormatDescriptionEvent
+	var found bool
+	onEventFunc := func(e *replication.BinlogEvent) error {
+		if e.Header.EventType != replication.FORMAT_DESCRIPTION_EVENT {
+			return errors.Errorf("got %+v, expect FormatDescriptionEvent", e.Header)
+		}
+		found = true
+		return nil
+	}
+
+	eof, err := replication.NewBinlogParser().ParseSingleEvent(f, onEventFunc)
+	if err != nil {
+		return false, errors.Annotatef(err, "parse %s", filename)
+	} else if eof {
+		return false, errors.Annotatef(io.EOF, "parse %s", filename)
+	}
+	return found, nil
+}
+
+// checkIsDuplicateEvent checks if the event is a duplicate event in the file.
+// NOTE: handle cases when file size > 4GB
+func checkIsDuplicateEvent(filename string, ev *replication.BinlogEvent) (bool, error) {
+	// 1. check event start/end pos with the file size
+	fs, err := os.Stat(filename)
+	if err != nil {
+		return false, errors.Annotatef(err, "get stat for %s", filename)
+	}
+	evStartPos := int64(ev.Header.LogPos - ev.Header.EventSize)
+	evEndPos := int64(ev.Header.LogPos)
+	if fs.Size() <= evStartPos {
+		return false, nil // the event not in the file
+	} else if fs.Size() < evEndPos {
+		// the file can not hold the whole event, often because the file is corrupt
+		return false, errors.Errorf("file size %d is between event's start pos (%d) and end pos (%d)",
```

Comment:
are there undetermined case, like flush partial binlog

Reply:
I think if one _fd_ write some data but not flushed other _fd_ still can see them, is it right? If it is, then this `if` should be `false`. 

NOTE: these file util functions are not concurrency-safe. 

if only flushed partial binlog event and this `if` is `true`, then we need to _recover_ the binlog file.

Reply:
if  is writing, it's possible. you should define to use it in non-concurrent scene

Reply:
add some comments for these util functions.